hngprojects/hng_boilerplate_nestjs
# feat: disable 2fa

## What does the PR do?

This PR is about disabling two factor authentication for logged in users. The endpoint returns errors if:

- [x] User is not logged in.
- [x] User does not exist.
- [x] Validation errors on the user input is done

How should this be manually tested?

1. Login to a user account, and get the access token

2. Make a POST request to the `api/v1/auth/2fa/disable` endpoint with the following payload:

```json
Authorization: Bearer <access_token>

{
    "password": "jetpack123",
    "totp_code": "123456"
}
```

Successful Response

On a successful testimonial creation, the API should return a 201 CREATED status code.

```json
{
    "status_code": 200,
    "message": "2FA is already disabled",
    "data": {
        "is_2fa_enabled": false
    }
}
```



Authentication Error:

When the current user is not properly authenticated to change the organization detail:

```json
{
  "status": "error",
  "message": "User is currently unauthorized, kindly authenticate to continue",
  "status_code": 401
}
```

Invalid UserID Error


```
![wrong user ID image](https://github.com/user-attachments/assets/6f9b6f4b-ff0d-4fa7-828f-aa61feff8440)

Validation:

If any required field is missing or invalid, the response should return a 400 Bad Request status code with appropriate validation error messages.

```json
{
  "status": "error",
  "message": ["name must be a string", "content must be a string"],
  "status_code": "400"
}
```

Task

- [x] Create a POST endpoint UPDATE /api/v1/2FA/disable to disable two factor authentication for logged in users.
- [x] Implement server-side logic to handle successful and failed update requests.
- [x] Ensure the user has the proper authentication to create testimonials.
- [x] Update the user two factor authenticator details in the database.

Reference Issue: [Link to Issue](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/266)## What does the PR do?

This PR is about authentication  for logged in users. The endpoint returns errors if:

- [x] User is not logged in.
- [x] User does not exist.
- [x] Validation errors on the user input is done

How should this be manually tested?

1. Login to a user account, and get the access token

2. Make a POST request to the `api/v1/auth/2fa/disable` endpoint with the following payload:

```json
Authorization: Bearer <access_token>
{
    "password": "jetpack123",
    "totp_code": "123456"
}
```

Successful Response

On a successful testimonial creation, the API should return a 201 CREATED status code.

```json
{
    "status_code": 200,
    "message": "2FA is already disabled",
    "data": {
        "is_2fa_enabled": false
    }
}
```
